cgen error: ')' expected (got "_t1")

### Describe the bug

```
==================
/tmp/v_1000/asd.11684854164238007274.tmp.c:12262: error: ')' expected (got "_t1")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Expected Behavior

for it to compile

### Current Behavior

error
```
==================
/tmp/v_1000/asd.11684854164238007274.tmp.c:12262: error: ')' expected (got "_t1")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```v
struct Foo {
pub:
	a int
}

fn foo() !Foo {
	return Foo{1}
}

fn main() {
	match foo()!.a {
		1 {
			println("1")
		}
		else {
			println("else")
		}
	}
}
```

Note: this version fixes it:
```v
struct Foo {
pub:
	a int
}

fn foo() !Foo {
	return Foo{1}
}

fn main() {
	a := foo()!.a
	match a {
		1 {
			println("1")
		}
		else {
			println("else")
		}
	}
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

latest: Current V version: V 0.3.2 0e727ec, timestamp: 2022-11-20 16:44:58 +0200

### Environment details (OS name and version, etc.)

OS: linux, Linux version 6.1.6-arch1-1 (linux@archlinux) (gcc (GCC) 12.2.0, GNU ld (GNU Binutils) 2.39.0) #1 SMP PREEMPT_DYNAMIC Sat, 14 Jan 2023 13:09:35 +0000
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-6400 CPU @ 2.70GHz
CC version: cc (GCC) 12.2.0

getwd: /home/asus/Projects/testbot
vmodules: /home/asus/.vmodules
vroot: /home/asus/.v
vexe: /home/asus/.v/v
vexe mtime: 2023-01-28 14:25:46
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 6e24f7e.d3e4058

Git version: git version 2.39.0
Git vroot status: weekly.2023.04-18-gd3e4058a
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -67,7 +67,9 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 	}
 	if node.cond in [ast.Ident, ast.IntegerLiteral, ast.StringLiteral, ast.FloatLiteral]
 		|| (node.cond is ast.SelectorExpr
-		&& (node.cond as ast.SelectorExpr).or_block.kind == .absent) {
+		&& (node.cond as ast.SelectorExpr).or_block.kind == .absent
+		&& ((node.cond as ast.SelectorExpr).expr !is ast.CallExpr
+		|| ((node.cond as ast.SelectorExpr).expr as ast.CallExpr).or_block.kind == .absent)) {
 		cond_var = g.expr_string(node.cond)
 	} else {
 		line := if is_expr {

--- a/vlib/v/tests/match_expr_with_opt_result_test.v
+++ b/vlib/v/tests/match_expr_with_opt_result_test.v
@@ -1,0 +1,32 @@
+struct Foo {
+pub:
+	a int
+}
+
+fn foo() !Foo {
+	return Foo{1}
+}
+
+fn bar() ?Foo {
+	return Foo{2}
+}
+
+fn test_main() {
+	match foo()!.a {
+		1 {
+			assert true
+		}
+		else {
+			assert false
+		}
+	}
+
+	match bar()?.a {
+		2 {
+			assert true
+		}
+		else {
+			assert false
+		}
+	}
+}

